File: euler/P16.java
// Project Euler, problem 16: Power digit sum (http://projecteuler.net/problem=16). 
// 215 = 32768 and the sum of its digits is 3 + 2 + 7 + 6 + 8 = 26.
// What is the sum of the digits of the number 21000?

import java.lang.Math; 
import java.math.BigInteger;

class P16 {
    
    public static void main(String[] args) {
	//Long isn't enough precise for such a big value. 
	//Double value is expressed in scientific notation. 
	String strNum = BigInteger.valueOf(2).pow(1000).toString();				
	long sum = 0; 

	for(int i = 0; i < strNum.length(); i++){	    
	    sum += Integer.parseInt(strNum.substring(i,i+1));
	}
	System.out.println("Sum: "+sum);		    	
    }
}

File: euler/p16.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

"""
Project Euler, problem 16: Power digit sum (http://projecteuler.net/problem=16).
2^15 = 32768 and the sum of its digits is 3 + 2 + 7 + 6 + 8 = 26.
What is the sum of the digits of the number 2^1000?
"""

def main():
    """
    The sum can be solved efficiently using map.
    """

    res = sum(map(int, list(str(2**1000))))

    print "The sum of the digits of the number 2^1000 is %d " % res

if __name__ == "__main__":
    main()

